compiler/optimizations/preNormalizeOptimizations.cpp

```diff
@@ -102,36 +117,256 @@ void doPreNormalizeArrayOptimizations() {
       if (fAutoLocalAccess) {
         autoLocalAccess(forall);
       }
+
+      if (fAutoAggregation) {
+        autoAggregation(forall);
+      }
+    }
+  }
+}
+
+Expr *preFoldMaybeLocalArrElem(CallExpr *call) {
+
+  // (call "may be local arr elem" elem, arr, paramFlag)
+  // we don't really care about the second argument here, it is used elsewhere
+
+  SymExpr *maybeArrElemSymExpr = toSymExpr(call->get(1));
+  INT_ASSERT(maybeArrElemSymExpr);
+
+  SymExpr *controlSymExp = toSymExpr(call->get(3));
+  INT_ASSERT(controlSymExp);
+
+  bool confirmed = (controlSymExp->symbol() == gTrue);
+
+  AggregationCandidateInfo *aggCandidate = preNormalizeAggCandidate[call];
```

Comment:
How can this map work under generic instantiation? Wouldn't you need to copy the map entry for calls an instantiated function? Seems to me it might be better to represent this info in the AST somehow so that it can be copied or removed naturally. PRIM_OPTIMIZATION_INFO could be used for this AFAIK.

Reply:
We populate this right before normalization, and use the information while preFolding calls. It was my understanding that we always preFold them before trying to resolve/instantiate them. Is that wrong? Then, yeah, I am not sure how this is working :)

Reply:
If you have code in a generic function, I'm pretty sure we copy (and substitute for instantiation) before preFold'ing the function body. But, I could be wrong about that. Could you look in to it?

Reply:
Yeah I remember having similar issues while implementing automatic local access. I'll look.

Reply:
You are right.

To use similar approaches to the rest of what we do, I am thinking of replacing the `=` with `PRIM_MAYBE_AGGREGATE_ASSIGN` that packs some of the information along with lhs and rhs of the assignment.

Reply:
That sounds workable to me. I think it'd be nice to have some more general strategies for this kind of thing (and was hoping PRIM_OPTIMIZATION_INFO would be that) but you don't have to use it for this.

Reply:
(Another option, besides primitives in the AST, is to add more fields to the ForallStmt class)

Reply:
.. and modify the `clone` function for ForallStmt, right?

Reply:
yes

Reply:
well, sortof. You would modify `ForallStmt::copyInner`.

Reply:
@mppf this is something we have discussed before. For this one, I stuck with the approach that I feel more comfortable with. But I created https://github.com/Cray/chapel-private/issues/1604 to record some related thoughts.